R/fifelse.R

```diff
@@ -0,0 +1,10 @@
+fifelse <- function(test, yes, no){
+  switch(class(yes)[1L],
```

Comment:
maybe best to use `is.factor`.

What's the expected behavior for something like `ifelse(TRUE, data.table(1:5), data.table(5:1))`? Same as for any other `list`? If so, should use `is.list` instead as well

Reply:
This is a tricky one! It is also related to the examples I mentioned in my post of [#3657 ](https://github.com/Rdatatable/data.table/issues/3657). 
I was asking for your thoughts given that it was not clear to me what should be returned for `list`?
  
You said:
_"I'm not sure how much attention we should devote to handling list output just yet, in any case._"

And I hence limited myself to implement something very basic.
  
As I mentioned `ifelse` does not always make sense to me.
`dplyr::if_else` and `htulis::if_else` are more strict, so is `fifelse`, but works sometimes differently.

**1/** Right now, here is what would be returned by fifelse:
  
```
fifelse(TRUE, data.table(1:5), data.table(5:1))
# Error in fifelse(test, yes, no) : Type list is not supported.

fifelse(TRUE, list(1:5), list(5:1))
# Error in fifelse_impl(test, unlist(yes), unlist(no)) : 
# Length of 'yes' and 'no' must be 1 or equal to length of 'test'. 

```
I believe the error message should be changed so that `data.table` is consistent with a `list`. the length of `yes` and `no` should be either equal to `1` or the length of `test`.
Anyway, `ifelse` would work for the above examples and `dplyr::if_else` would only work for 
the second example.

**2/** Note that the following would work:
  
```
fifelse(-2L:2L < 0, list(1:5), list(5:1))
# [[1]]
# [1] 1
# 
# [[2]]
# [1] 2
# 
# [[3]]
# [1] 3
# 
# [[4]]
# [1] 2
# 
# [[5]]
# [1] 1
```
but would return something different than `ifelse` and `dyplr::if_else`.
I was thinking about adding support to data.frame and data.table and treat them like a `list`,
what do you think?

If you refer to my post on Issue [#3657 ](https://github.com/Rdatatable/data.table/issues/3657) you will find more examples with `list`, for which
I am not really sure what to return. As shown ifelse returns (in some cases) outputs that I do not
understand or outputs that loose the class type.

In any cases, I am happy to implement what the members of the Rdatatable team believe make sense. Please do let me know what you think the fifelse function should return in these examples (including [#3657 ](https://github.com/Rdatatable/data.table/issues/3657) ). It is quite difficult to say what is right or wrong here.

As to your suggestion to use `is.factor` and `is.list`, I believe that would involve more test and `if` functions and would slow down the code compare to a single `switch` test that does a dispatch to the  corresponding value. I may be wrong, if so, please feel free to suggest an alternative implementation and I would not mind at all amending the code.

Reply:
indeed... this rabbit hole is why I originally suggested to keep away. Actually I'd expect `fifelse(TRUE, list(1:5), list(5:1))` to return `list(1:5)` -- this is how `ifelse` would be used to replace elements of a list column, for example.

I would stick to using `if (is.factor)` and `if (is.list)` branches, with the `list` branch requiring `(length(yes) == length(test) || length(yes) == 1L) && (length(no) == length(test) || length(no) == 1L)`. The efficiency difference (inasmuch as there is one, which I'm not sure of) will be on the order of nanoseconds, and only happens once per function call.

Reply:
i.e. I would expect the following normal usage of `fifelse`:

```
DT = data.table(a = 0:5, b = list(1, list(2:3), list(4:6), list(6:4), list(3:2), 1))
DT[ , b := fifelse(a > 2, b, list(0))]
#now, DT$b == list(0, 0, 0, list(6:4), list(3:2), 1), i.e. rows 4,5,6 unaffected
```

List columns are a bit funny... if I'm not mistaken, it would also be consistent (and convenient) if singleton replacements could be en-listed (wrapped correctly in `list()`), i.e.:

```
DT = data.table(a = 0:5, b = list(1, list(2:3), list(4:6), list(6:4), list(3:2), 1))
DT[ , b := fifelse(a > 2, b, 0)]
```

Could also work (a bit of an anti-pattern but I think the intent is clear). This is consistent with `list` assignment directly:

```
DT = data.table(a = list(1))
DT[1, a := 0]
# now DT$a == list(0)
```

Reply:
I have pushed a new version. Let's see if it passes tests. 
Your examples should all work now.
`is.list` is not needed anymore as list are now handle directly by the C code.
This test `(length(yes) == length(test) || length(yes) == 1L) && (length(no) == length(test) || length(no) == 1L)` is not needed either as something similar is done in the C code.
More tests have been added to `test.Rraw` file.
When I have time, I will be looking at moving the rest of R code to C (if possible) so `is.factor` or `class` are not needed. I will be also looking at the factor/character topic mentioned by @jangorecki .

Thanks for the feedback.